sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEUtils.scala

```diff
@@ -28,16 +28,31 @@ object AQEUtils {
   def getRequiredDistribution(p: SparkPlan): Option[Distribution] = p match {
     // User-specified repartition is only effective when it's the root node, or under
     // Project/Filter/LocalSort/CollectMetrics.
-    // Note: we only care about `HashPartitioning` as `EnsureRequirements` can only optimize out
-    // user-specified repartition with `HashPartitioning`.
-    case ShuffleExchangeExec(h: HashPartitioning, _, shuffleOrigin)
+    // Note, here are two cases of how user-specified repartition can be optimized out:
+    // 1. `EnsureRequirements` can only optimize out user-specified repartition with
+    //    `HashPartitioning`.
+    // 2. `AQEOptimizer` can optimize out user-specified repartition with all `Partitioning`,
+    //     e.g. convert empty to local relation.
```

Comment:
shall we fix `PropagateEmptyRelationBase` instead? I don't think we can optimize out `Repartition` which breaks user expectations. The change here only covers AQE and I think this is a problem for non AQE as well.

Reply:
Or, we can still optimize out `Repartition`, but we should assign a `Partitioning` to `LocalRelation`. Ideally a empty data can be any partitioning.

Reply:
> shall we fix PropagateEmptyRelationBase instead? I don't think we can optimize out Repartition which breaks user expectations. The change here only covers AQE and I think this is a problem for non AQE as well.

if we do not optimize the repartition at the top of empty relation, we also can not optimize other plan at the top of repartition. It may become a regression. We should only preserve the final repatition which can affect the final output partition as the `requiredDistribution` did.

> Or, we can still optimize out Repartition, but we should assign a Partitioning to LocalRelation. Ideally a empty data can be any partitioning.

I actually thought about it, but it's not easy. If we want to assign a `Partitioning` to `LocalRelation`, we also need to consider how to propagate the `Partitioning` when we through other plan otherwise it does less meaning. Then we need care the output partitioning for all logical plan. This should not be expected since the partitioning is a physical concept.

Reply:
OK let me make my proposal clear: let's not optimize out repartition if it's the root node, or below Project/Filter, in any cases. What do you think?

Reply:
skip optimize root user-specified repartition by https://github.com/apache/spark/pull/37612/commits/546628906f0329f523fc26ead4913f3032ec9851